tests/python/relay/test_dataflow_pattern.py

```diff
@@ -0,0 +1,853 @@
+
+def test_AltPattern():
+    is_add_or_sub = is_op('add') | is_op('subtract')
+    assert isinstance(is_add_or_sub, AltPattern)
+
+def test_TypePattern():
+    ttype = relay.TensorType((10, 10), "float32")
+    ty_pat = has_type(ttype)
+    assert isinstance(ty_pat, TypePattern)
+    assert ty_pat.type == ttype
+
+def test_AttrPattern():
+    op = is_op('add').has_attr("TOpPattern", K_ELEMWISE)
+    assert isinstance(op, AttrPattern)
+    assert op.attrs["TOpPattern"] == K_ELEMWISE
+
+## MATCHER TESTS
+
+def test_match_op():
+    assert is_op('add').match(relay.op.op.get("add"))
+
+def test_no_match_op():
+    assert not is_op('add').match(relay.op.op.get("subtract"))
+
+def test_match_op_or():
+    is_add_or_sub = is_op('add') | is_op('subtract')
+    assert is_add_or_sub.match(relay.op.op.get("add"))
+    assert is_add_or_sub.match(relay.op.op.get("subtract"))
+
+def test_match_call_commutive():
+    x = relay.var('x')
+    y = relay.var('y')
+    add_pattern = is_op('add')(is_input("x"), is_input("y"))
+    assert add_pattern.match(x + y)
+    assert add_pattern.match(y + x)
+    mul_pattern = is_op('multiply')(is_input("x"), is_input("y"))
+    assert mul_pattern.match(x * y)
+    assert mul_pattern.match(y * x)
+
+def test_no_match_call_commutive():
+    x = relay.var('x')
+    y = relay.var('y')
+    add_pattern = is_op('subtract')(is_input("x"), is_input("y"))
+    assert add_pattern.match(x - y)
+    assert not add_pattern.match(y - x)
+    add_pattern = is_op('divide')(is_input("x"), is_input("y"))
+    assert add_pattern.match(x / y)
+    assert not add_pattern.match(y / x)
+
+def test_match_call():
+    x = relay.var('x')
+    y = relay.var('y')
+    add_pattern = is_op('add')(wildcard(), wildcard())
+    assert add_pattern.match(x + y)
+
+def test_no_match_call():
+    x = relay.var('x')
+    y = relay.var('y')
+    add_pattern = is_op('add')(wildcard(), wildcard())
+    assert not add_pattern.match(x - y)
+
+def test_match_option():
+    x = relay.var('x')
+    w = relay.var('w')
+    b = relay.var('b')
+    pattern = is_op("nn.relu")(
+            is_op("nn.conv2d")(wildcard(), wildcard()
+                             ).optional(lambda x: is_op("nn.bias_add")(x, wildcard()))
+        )
+
+    conv2d = relay.op.nn.conv2d(x, w)
+    relu = relay.op.nn.relu(conv2d)
+    assert pattern.match(relu)
+
+    conv2d = relay.op.nn.conv2d(x, w)
+    bias_add = relay.op.nn.bias_add(conv2d, b)
+    relu = relay.op.nn.relu(bias_add)
+    assert pattern.match(relu)
+
+    pattern = is_op("nn.conv2d")(wildcard(), wildcard())
+    pattern = pattern.optional(is_op('nn.relu')).optional(is_op("tanh"))
+
+    conv2d = relay.op.nn.conv2d(x, w)
+    relu = relay.op.nn.relu(conv2d)
+    tanh = relay.op.tanh(conv2d)
+    tanh2 = relay.op.tanh(relu)
+    relu2 = relay.op.nn.relu(tanh)
+    assert pattern.match(conv2d)
+    assert pattern.match(relu)
+    assert pattern.match(tanh)
+    assert pattern.match(tanh2)
+    assert not pattern.match(relu2)
```

Comment:
Should it really fail? I think conv2d -> tanh -> relu should partially match this pattern. I'd be surprised if `conv2d -> tanh` matches but `conv2d -> tanh -> relu` does not.

Reply:
Is this what you mean by `it still imposes the order you construct them in`?

Reply:
This is the order issue I mentioned. The pattern is `conv2d -> ?relu -> ?tanh`

`conv2d -> tanh` 
In this one, tanh matches ?tanh
conv2d does not match ?relu
conv2d does match conv2d, pattern matches

`conv2d -> tanh -> relu`
In this one, relu does not match ?tanh
it does match ?relu, so we try to match the tanh 
tanh does not match conv2d, 
there are no other options, the match fails

This pattern would match:
`conv2d -> (?relu -> ?tanh | ?tanh -> ?relu)`

Reply:
ok, as long as there is no consecutive optionals, matching would work as expected, right? I think it is very reasonable trade off in terms of benefit / implementation complexity.

Reply:
Yep, same matching rules as usual, just a slight opening for "you can skip this one if you need to"